File: src/Logging.cpp
#include "wx/datetime.h"
#include "wx/filename.h"

#include "Logging.h"


Logging::Logging(const wxString &t_log_file_name)
    :
        m_old_logger(wxLog::GetActiveTarget()),
        m_log_file(t_log_file_name, "w+")
    {
    wxLog::SetActiveTarget(this);

    if (is_ok()) {
        auto session_str = wxString::Format("Session started at %s\n", wxDateTime::Now().FormatISODate());
        m_log_file.Write(session_str.c_str(), session_str.length());
        m_log_file.Sync();
    }
    wxLogInfo("=== LOGGING STARTED ===");
}

Logging::~Logging() {
    wxLogInfo("=== LOGGING FINISHED ===");
    m_log_file.Close();
    wxLog::SetActiveTarget(m_old_logger);
}


void Logging::DoLogLine(const wxString &t_level,
                        const wxString &t_timestr,
                        const wxString &t_threadstr,
                        const wxString &t_msg) {
    if (is_ok()) {
        auto t_str = wxString::Format("<%8s> %8s [%10s] %s", t_level, t_timestr, t_threadstr, t_msg);
        m_log_file.Write(t_str.c_str(), t_str.length());
        m_log_file.Sync();
    }
}

void Logging::DoLogRecord(wxLogLevel t_level,
                          const wxString &t_msg,
                          const wxLogRecordInfo &t_info) {
    wxString level_str;
    wxString postfix = "";

    if (t_level <= wxLOG_Warning || t_level == wxLOG_Trace || t_level == wxLOG_Debug) {
        postfix = wxString::Format(" [%s() in %s(%d)]", t_info.func, wxFileName(t_info.filename).GetFullName(), t_info.line);
        m_old_logger->LogRecord(t_level, t_msg, t_info);
    }

    switch (t_level) {
        case wxLOG_Error:
            level_str = "ERROR";
            break;

        case wxLOG_Warning:
            level_str = "WARNING";
            break;

        case wxLOG_Message:
            level_str = "MESSAGE";
            break;

        case wxLOG_Status:
            level_str = "STATUS";
            break;

        case wxLOG_Info:
            level_str = "INFO";
            break;

        case wxLOG_Debug: 
            level_str = "DEBUG";
            break;

        case wxLOG_Trace:
            level_str = "TRACE";
            break;

        case wxLOG_Progress:
            level_str = "PROGRESS";
            break;

        default:
            level_str = "USER";
    }

    DoLogLine(
        level_str,
        wxDateTime(wxLongLong(t_info.timestampMS)).FormatISOTime(),
        t_info.threadId == wxThread::GetMainId()
            ? wxString("main")
            : wxString::Format("%lx", t_info.threadId),
        wxString::Format("%s%s\n", t_msg, postfix)
    );
}

File: Img2STL/utils/logging.py
import os
import wx

from datetime import datetime


class Logging(wx.Log):
    def __init__(self, log_file_name: str) -> None:
        super().__init__()
        self._old_logger: wx.Log = wx.Log.GetActiveTarget()
        self._log_file_name = os.path.join(os.getcwd(),  log_file_name)

        wx.Log.SetActiveTarget(self)

        with open(self._log_file_name, "w+") as flog:
            flog.write(f"Session started at {wx.DateTime.Now().FormatISODate()}\n")

    def __del__(self) -> None:
        wx.Log.SetActiveTarget(self._old_logger)

    def do_log_line(self, level: str, timestr: str, msg: str) -> None:
        with open(self._log_file_name, "a+") as flog:
            t_str = f"<{level:8}> {timestr:8} {msg}\n"
            flog.write(t_str)

    def DoLogRecord(self, level, msg: str, info: wx.LogRecordInfo) -> None:
        postfix = ""

        if level <= wx.LOG_Warning or level == wx.LOG_Trace or level == wx.LOG_Debug:
            postfix = f"[{info.func}() in {info.filename}({info.line})]"
            self._old_logger.LogRecord(level, f"{level}: {msg}", info)

        if level == wx.LOG_Error:
            level_str = "ERROR"
        elif level == wx.LOG_Warning:
            level_str = "WARNING"
        elif level == wx.LOG_Message:
            level_str = "MESSAGE"
        elif level == wx.LOG_Status:
            level_str = "STATUS"
        elif level == wx.LOG_Info:
            level_str = "INFO"
        elif level == wx.LOG_Debug: 
            level_str = "DEBUG"
        elif level == wx.LOG_Trace:
            level_str = "TRACE"
        elif level == wx.LOG_Progress:
            level_str = "PROGRESS"
        else:
            level_str = "USER"

        self.do_log_line(
            level_str,
            datetime.utcfromtimestamp(info.timestamp).strftime("%H:%M:%S"),
            f"{msg} {postfix}"
        )
